src/hooks/useArray.ts

```diff
@@ -30,18 +31,25 @@ export type UseArrayResult<T> = {
   isDirty: boolean;
   setDirty: () => void;
   clearDirty: () => void;
+  replace: UseArrayReplace<T>;
 };
 
 export const useArray = <T>(
   array?: T[],
   options?: UseArrayOptions<T>
 ): UseArrayResult<T> => {
   const create = options?.create ?? createPlainObject;
-  const prepared = options?.prepare ? array?.map(options.prepare) : array;
+  const [prepared, setPrepared] = useState(
+    options?.prepare ? array?.map(options.prepare) : array
+  );
 
   const [items, setItems] = useState(prepared ?? []);
-  const [isDirty, setIsDirty] = useState(false);
 
+  useEffect(() => {
+    setItems(prepared ?? []);
+  }, [prepared]);
```

Comment:
My take on this:

No side effects inside `useArray`, this will make it do unexpected things. Have the appropriate array operation be called inside `useEffect()` inside a component, that makes it clear that it's part of the component lifecycle, not hidden mutation of the array data that may or may not trigger component re-render.

Also, the above is a comment on first view, still don't see the need for all of this.